mbedtls/src/ssl/async_session.rs

```diff
@@ -130,9 +130,81 @@ impl Drop for SessionWrapper<'_> {
     }
 }
 
+// mbedtls_ssl_write() has some weird semantics w.r.t non-blocking I/O:
+//
+// > When this function returns MBEDTLS_ERR_SSL_WANT_WRITE/READ, it must be
+// > called later **with the same arguments**, until it returns a value greater
+// > than or equal to 0. When the function returns MBEDTLS_ERR_SSL_WANT_WRITE
+// > there may be some partial data in the output buffer, however this is not
+// > yet sent.
+//
+// WriteTracker is used to ensure we pass the same data in that scenario.
+//
+// Reference:
+// https://tls.mbed.org/api/ssl_8h.html#a5bbda87d484de82df730758b475f32e5
+struct WriteTracker {
+    pending: Option<Box<DigestAndLen>>,
+}
+
+struct DigestAndLen {
+    digest: [u8; 20], // SHA-1
+    len: usize,
+}
+
+impl WriteTracker {
+    fn new() -> Self {
+        WriteTracker {
+            pending: None,
+        }
+    }
+
+    fn digest(buf: &[u8]) -> [u8; 20] {
+        use crate::hash::{Md, Type};
+        let mut out = [0u8; 20];
+        let res = Md::hash(Type::Sha1, buf, &mut out[..]);
+        assert_eq!(res, Ok(out.len()));
+        out
+    }
+
+    fn adjust_buf<'a>(&self, buf: &'a [u8]) -> io::Result<&'a [u8]> {
+        match self.pending.as_ref() {
+            None => Ok(buf),
+            Some(pending) => {
+                if pending.len <= buf.len() {
+                    let buf = &buf[..pending.len];
+                    if Self::digest(buf) == pending.digest {
```

Comment:
Do we really need to do this? It's expensive and unlikely to occur.

Reply:
I agree that it is unlikely but if it happens it can cause a lot of trouble. Alternative would be to assume that the same bytes (with maybe more bytes appended) are written and just make sure we slice it to the previous length. And of course document the behavior, but we won't be able to throw an error if the conditions are not met. Do you think that's worth it?

Reply:
Can't we do both? Do the expensive check when the crate is compiled in debug mode, and assume everything is correct in release mode?

Reply:
I like that idea.

Reply:
Taking the liberty of resolving this (was implemented, see `debug_assertions`)